sound/soc/sof/ipc3-control.c

```diff
@@ -96,6 +96,28 @@ static int sof_ipc3_set_get_kcontrol_data(struct snd_sof_control *scontrol,
 	cdata->elems_remaining = 0;
 
 	ret = iops->set_get_data(sdev, cdata, cdata->rhdr.hdr.size, set);
+	if (!set)
+		goto unlock;
+
+	/* It is a set-data operation, and we have a backup that we can restore */
+	if (ret < 0) {
+		if (scontrol->ipc_control_data_valid)
+			goto unlock;
+		/* Current ipc_control_data is not valid, we use the last known good
+		 * configuration
+		 */
+		memcpy(scontrol->ipc_control_data, scontrol->old_ipc_control_data,
+		       scontrol->max_size);
```

Comment:
The memory is preallocated from the time the entire structure is allocated. We can extend this kind of backup/delayed validation logic to other types of data indeed, if you think that's a good idea I can do it too.

Reply:
I don't think we would have a blanking out, since a backup... currently isn't made in the first place (so the restore logic doesn't trigger because the ipc_data_valid flag never changes to false); to opt in for that control you make the copy and then set the flag to false.

This is done per kcontrol, individually.